wp-content/plugins/goonj-blocks/build/functions.php

```diff
@@ -15,4 +15,208 @@ function gb_format_time_range($start_date, $end_date, $format = 'h:i A') {
     }
     return $start_date->format($format) . ' - ' . $end_date->format($format);
-}
\ No newline at end of file
+}
+/**
+ * Generate induction slots based on contact's location, registration date, and induction activity details.
+ * 
+ * @param int|null $contactId The ID of the contact for whom to generate slots.
+ * @param int $days The number of days to look ahead for slot generation (default is 30 days).
+ * @return array An array of induction slots or empty if criteria aren't met.
+ */
+function generate_induction_slots($contactId = null, $days = 30) {
+    // Validate contactId
+    if (empty($contactId) || !is_numeric($contactId)) {
+        \Civi::log()->warning('Invalid contactId', ['contactId' => $contactId]);
+        return [];
+    }
+
+    try {
+        // Fetch necessary contact details for slot generation
+        $contactData = \Civi\Api4\Contact::get(FALSE)
+            ->addSelect('address_primary.state_province_id', 'address_primary.city', 'Individual_fields.Created_Date')
+            ->addWhere('id', '=', $contactId)
+            ->execute()->single();
+
+        if (empty($contactData)) {
+            \Civi::log()->warning('Contact not found', ['contactId' => $contactId]);
+            return [];
+        }
+
+        // Retrieve induction activity for the contact
+        $inductionActivity = \Civi\Api4\Activity::get(FALSE)
+            ->addSelect('id', 'activity_date_time', 'status_id', 'status_id:name', 'Induction_Fields.Goonj_Office')
+            ->addWhere('source_contact_id', '=', $contactId)
+            ->addWhere('activity_type_id:name', '=', 'Induction')
+            ->execute()->single();
+
+        // Return if no induction activity found
+        if (empty($inductionActivity)) {
+            \Civi::log()->info('No activity found for contact', ['contactId' => $contactId]);
+            return [];
+        }
+
+        // If the induction activity status is 'Scheduled' or 'Completed', return the details
+
+        if (in_array($inductionActivity['status_id:name'], ['Scheduled', 'Completed'])) {
+            return [
+                'status' => $inductionActivity['status_id:name'],
+                'date' => (new DateTime($inductionActivity['activity_date_time']))->format('d-m-Y H:i')
+            ];
+        }
+
+        $contactStateId = intval($contactData['address_primary.state_province_id']);
+        $inductionSlotStartDate = (new DateTime($contactData['Individual_fields.Created_Date']))->modify('+1 day');
+        $physicalInductionType = 'Processing_Unit';
+        $onlineInductionType = 'Online_only_selected_by_Urban_P';
+        $defaultMaxSlot = 30;
+
+        // List of states with both physical and online inductions based on cities
+        $statesWithMixedInductionTypes = \Civi\Api4\StateProvince::get(FALSE)
+            ->addWhere('country_id.name', '=', 'India')
+            ->addWhere('name', 'IN', ['Bihar', 'Jharkhand', 'Orissa'])
```

Comment:
If we are setting the addWhere here and fetching 3 states, should we need to add this line `->setLimit(3)`, can you please remove ->setLimit(3) it and then test it working fine or not

Reply:
removed